public/app/core/components/RolePicker/RolePickerMenu.tsx

```diff
@@ -0,0 +1,585 @@
+  const [selectedBuiltInRole, setSelectedBuiltInRole] = useState<OrgRole>(builtInRole);
+  const [selectedBuiltInRoles, setSelectedBuiltInRoles] = useState<Role[]>(builtInRoles[builtInRole]);
+  const [showSubMenu, setShowSubMenu] = useState(false);
+  const [openedMenuGroup, setOpenedMenuGroup] = useState('');
+  const [subMenuOptions, setSubMenuOptions] = useState<Role[]>([]);
+
+  const theme = useTheme2();
+  const styles = getSelectStyles(theme);
+  const customStyles = useStyles2(getStyles);
+
+  const getOptionGroups = useCallback(() => {
+    const groupsMap: { [key: string]: Role[] } = {};
+    options.forEach((role) => {
+      if (role.name.startsWith('fixed:')) {
+        const groupName = getRoleGroup(role);
+        if (groupsMap[groupName]) {
+          groupsMap[groupName].push(role);
+        } else {
+          groupsMap[groupName] = [role];
+        }
+      }
+    });
+
+    const groups = [];
+    for (const groupName in groupsMap) {
+      if (Object.prototype.hasOwnProperty.call(groupsMap, groupName)) {
+        const groupOptions = groupsMap[groupName].sort(sortRolesByName);
+        groups.push({
+          name: fixedRoleGroupNames[groupName] || capitalize(groupName),
+          value: groupName,
+          options: groupOptions,
+        });
+      }
+    }
+    return groups.sort((a, b) => (a.name < b.name ? -1 : 1));
+  }, [options]);
+
+  // Call onSelect() on every selectedOptions change
+  useEffect(() => {
+    onSelect(selectedOptions);
+  }, [selectedOptions, onSelect]);
+
+  useEffect(() => {
+    if (onBuiltInRoleSelect) {
+      onBuiltInRoleSelect(selectedBuiltInRole);
+    }
+  }, [selectedBuiltInRole, onBuiltInRoleSelect]);
+
+  const customRoles = options.filter(filterCustomRoles).sort(sortRolesByName);
+  const fixedRoles = options.filter(filterFixedRoles).sort(sortRolesByName);
+  const optionGroups = getOptionGroups();
+
+  const getSelectedGroupOptions = (group: string) => {
+    const selectedGroupOptions = [];
+    for (const role of selectedOptions) {
+      if (getRoleGroup(role) === group) {
+        selectedGroupOptions.push(role);
+      }
+    }
+    for (const role of selectedBuiltInRoles) {
+      if (getRoleGroup(role) === group) {
+        selectedGroupOptions.push(role);
+      }
+    }
+    return selectedGroupOptions;
+  };
+
+  const groupSelected = (group: string) => {
+    const selectedGroupOptions = getSelectedGroupOptions(group);
+    const groupOptions = optionGroups.find((g) => g.value === group);
+    return selectedGroupOptions.length > 0 && selectedGroupOptions.length >= groupOptions!.options.length;
+  };
+
+  const groupPartiallySelected = (group: string) => {
+    const selectedGroupOptions = getSelectedGroupOptions(group);
+    const groupOptions = optionGroups.find((g) => g.value === group);
+    return selectedGroupOptions.length > 0 && selectedGroupOptions.length < groupOptions!.options.length;
+  };
```

Comment:
There's a fair bit of duplication between these two functions, maybe they can be simplified somehow?